ntplib.py

```diff
@@ -297,37 +300,62 @@ def request(self, host, version=2, port="ntp", timeout=5, address_family=socket.
         NTPStats object
         """
 
-        # lookup server address
-        addrinfo = socket.getaddrinfo(host, port, address_family)[0]
-        family, sockaddr = addrinfo[0], addrinfo[4]
-
-        # create the socket
-        sock = socket.socket(family, socket.SOCK_DGRAM)
-
+        dest_timestamp = None
         try:
-            sock.settimeout(timeout)
-
-            # create the request packet - mode 3 is client
-            query_packet = NTPPacket(
-                mode=3,
-                version=version,
-                tx_timestamp=system_to_ntp_time(time.time())
-            )
-
-            # send the request
-            sock.sendto(query_packet.to_data(), sockaddr)
-
-            # wait for the response - check the source address
-            src_addr = (None,)
-            while src_addr[0] != sockaddr[0]:
-                response_packet, src_addr = sock.recvfrom(256)
-
-            # build the destination timestamp
-            dest_timestamp = system_to_ntp_time(time.time())
-        except socket.timeout:
-            raise NTPException("No response received from %s." % host)
-        finally:
-            sock.close()
+            # lookup server address - request only datagram (i.e. UDP) addresses
+            addrinfo_a = socket.getaddrinfo(host, port, address_family, socket.SOCK_DGRAM)
+        except socket.gaierror:
+            # can happen for a v6 only resolvable hostname on a v4 only host (or a v4 only hostname on a v6 only host)
+            # getaddrinfo() tries to be clever; which is damn annoying!
+            raise NTPException("No resolved address(es) for %s." % host)
+
+        # we randomize the addresses in order to spread the load
+        # if a specific address does not work, no worries, we just continue to the next
+        # for sanity reasons (i.e. limiting random sample cpu usage); we limit the tries to _MAX_HOSTS_TO_TRY addresses
+        for addrinfo in random.sample(addrinfo_a, k=min(NTPClient._MAX_HOSTS_TO_TRY, len(addrinfo_a))):
+            if addrinfo[1] != socket.SOCK_DGRAM or addrinfo[2] != socket.IPPROTO_UDP:
```

Comment:
I don't think this should happen if we pass `SOCK_DGRAM` to `getaddrinfo` above, so might as well remove.

Reply:
This is me defaulting to "belt and suspenders"[1] style coding. In this case it's all about documenting and coding the correct path thru the code. I agree it's not 100% needed; however, it does force the code into never hitting a failure mode.

[1] Definition: "Involving or employing multiple methods or procedures to achieve a desired result especially out of caution or fear of failure"